src/compiler/checker.ts

```diff
@@ -23653,6 +23653,26 @@ namespace ts {
                 return type;
             }
 
+            function narrowTypeByPrivateIdentifierInInExpression(type: Type, expr: PrivateIdentifierInInExpression, assumeTrue: boolean): Type {
+                const target = getReferenceCandidate(expr.expression);
+                if (!isMatchingReference(reference, target)) {
+                    return type;
+                }
+
+                const privateId = expr.name;
+                const symbol = lookupSymbolForPrivateIdentifierDeclaration(privateId.escapedText, privateId);
+                if (symbol === undefined) {
+                    return type;
+                }
+                const classSymbol = symbol.parent!;
+                const classDecl = symbol.valueDeclaration;
+                Debug.assert(classDecl, "should always have a declaration");
+                const targetType = hasStaticModifier(classDecl)

```

Comment:
classDecl should be named privateDecl. Actually I would personally inline it like so, but it's up to you:
```suggestion
                const targetType = hasStaticModifier(Debug.checkDefined(symbol.valueDeclaration, "should always have a declaration"))
```

Reply:
done :)